src/ModbusSlave.cpp

```diff
@@ -462,31 +462,36 @@ uint8_t Modbus::writeRegisterToBuffer(int offset, uint16_t value)
     }
 
     _responseBuffer[index] = value >> 8;
-    _responseBuffer[index + 1] = value & 0xff;
+    _responseBuffer[index + 1] = value & 0xFF;
 
     return STATUS_OK;
 }
 
 /**
  * Writes an uint8_t array to the output buffer.
  *
- * @param offset The offset from the first register in the buffer.
- * @param str The array to write into the buffer.
- * @param length The array length.
+ * @param offset The offset from the first data register in the response buffer.
+ * @param str The array to write into the response buffer.
+ * @param length The length of the array.
  * @return STATUS_OK if succeeded, STATUS_ILLEGAL_DATA_ADDRESS if the data doesn't fit in the buffer.
  */
-uint8_t Modbus::writeStringToBuffer(int offset, uint8_t *str, uint8_t length)
+uint8_t Modbus::writeArrayToBuffer(int offset, uint16_t *str, uint8_t length)
 {
-    // (1 x valueBytes, n x values).
+    // Index to start writing from (1 x valueBytes, n x values (offset)).
     uint8_t index = MODBUS_DATA_INDEX + 1 + (offset * 2);
 
-    // Check the length of the array.
-    if ((index + length) > _responseBufferLength - MODBUS_CRC_LENGTH)
+    // Check if the array fits in the remaining space of the response.
+    if ((index + (length * 2)) > _responseBufferLength - MODBUS_CRC_LENGTH)
     {
+        // If not return an exception.
         return STATUS_ILLEGAL_DATA_ADDRESS;
     }
 
-    memcpy(_responseBuffer + index, str, length);
+    for (int i = 0; i < length; i++)
```

Comment:
hmm, a loop instead of memcpy ? are you sure ...
can you explain in a remark what is happening and continue useing memcpy ?

Reply:
This is done to enable passing a uint16_t array. As far as I know there is no suitable way to split a uint16_t into two uint8_t's and copy it to an array via memcpy.

Reply:
where does the endiness comes from ?
if you consider using `memcpy` vs `loop`, the difference is the endiness,
is it a problem you can solve on generating side ?
 
``` c
#include<stdio.h>
#include<stdint.h>
#include<string.h>

void main() {
    int length = 5;
    uint8_t _responseBuffer[] = {0,0,0,0,0,0,0,0,0,0};
    uint16_t str[] = {(5 << 8) + 1, (4 << 8) + 2,(3 << 8) + 3, (2 << 8) + 4, (1 << 8) + 5};

    printf ("copy using loop\n");
    for (int i = 0; i < length; i++)
    {
        _responseBuffer[(i * 2)] = str[i] >> 8;
        _responseBuffer[(i * 2) + 1] = str[i] & 0xFF;
    }
    for (int i = 0; i < length; i++)
    {
        printf ("%d, %d\n",
            _responseBuffer[(i * 2)],
            _responseBuffer[(i * 2) + 1]);
    }

    printf ("copy using memcpy\n");
    memcpy((uint8_t*)_responseBuffer, str, length * 2);
    for (int i = 0; i < length; i++)
    {
        printf ("%d, %d\n",
            _responseBuffer[(i * 2)],
            _responseBuffer[(i * 2) + 1]);
    }

    return;
}
```

Reply:
I didn't know that was possible with memcpy, but when running your example it seems to change the endianess of the values. Modbus expects a Big endian for the 16 bits values, and when using memcpy it switches it to little endian.

I'm also not fond of using a loop, but it seems (to me) to be the best readable and predictable option.

Reply:
ok :-)

actually, I just remember why this method exist :-)
it was made to store r/w **stings**, like maintenance dates, versions, address of device, etc...
the idea was that users could store strings in the registers space :rainbow:

Reply:
That would explain the name and could be a useful application. Having both would then be the ideal way.

Reply:
The importance of good method comments ... :palm_tree: :fox_face: